_posts/2021-12-3-arrow-duckdb.md

```diff
@@ -0,0 +1,383 @@
+ - https://github.com/duckdb/duckdb-web/blob/master/_posts/data/integers.parquet?raw=true
+ - https://github.com/cwida/duckdb-data/releases/download/v1.0/lineitemsf1.snappy.parquet
+
+
+#### Python
+
+There are two ways in Python of querying data from Arrow:
+1. Through the Relational API
+```py
+# Reads Parquet File to an Arrow Table
+arrow_table = pq.read_table('integers.parquet')
+
+# Transforms Arrow Table -> DuckDB Relation
+rel_from_arrow = duckdb.arrow(arrow_table)
+
+# we can run a SQL query on this and print the result
+print(rel_from_arrow.query('arrow_table', 'SELECT SUM(data) FROM arrow_table WHERE data > 50').fetchone())
+
+# Transforms DuckDB Relation -> Arrow Table
+arrow_table_from_duckdb = rel_from_arrow.arrow()
+```
+
+2. By using replacement scans and querying the object directly with SQL:
+```py
+# Reads Parquet File to an Arrow Table
+arrow_table = pq.read_table('integers.parquet')
+
+# Gets Database Connection
+con = duckdb.connect()
+
+# we can run a SQL query on this and print the result
+print(con.execute('SELECT SUM(data) FROM arrow_table WHERE data > 50').fetchone())
+
+# Transforms Query Result from DuckDB to Arrow Table
+# We can directly read the arrow object through DuckDB's replacement scans.
+con.execute("SELECT * FROM arrow_table").fetch_arrow_table()
+```
+
+It is possible to transform both DuckDB Relations and Query Results back to Arrow.
+
+#### R
+
+In R, the primary way to interact with Arrow is by registering the table as a view (an alternative is to use dplyr as shown above).
+```r
+library(duckdb)
+library(arrow)
+library(dplyr)
+
+# Reads Parquet File to an Arrow Table
+arrow_table <- arrow::read_parquet("integers.parquet", as_data_frame = FALSE)
+
+# Gets Database Connection
+con <- dbConnect(duckdb::duckdb())
+
+# Registers arrow table as a DuckDB view
+arrow::to_duckdb(arrow_table, table_name = "arrow_table", con = con)
+
+# we can run a SQL query on this and print the result
+print(dbGetQuery(con, "SELECT SUM(data) FROM arrow_table WHERE data > 50"))
+
+# Transforms Query Result from DuckDB to Arrow Table
+result <- dbSendQuery(con, "SELECT * FROM arrow_table")
+```
+
+### Streaming Data from/to Arrow
+In the previous section, we depicted how to interact with Arrow tables. However, Arrow also allows users to interact with the data in a streaming fashion. Either consuming it (e.g., from an Arrow Dataset) or producing it (e.g., returning a RecordBatchReader). And of course, DuckDB is able to consume Datasets and produce RecordBatchReaders. This example uses the NYC Taxi Dataset, stored in Parquet files partitioned by year and month, which we can download through the Arrow R package:
+```R
+arrow::copy_files("s3://ursa-labs-taxi-data", "nyc-taxi")
+```
+
+
+#### Python
+```py
+# Reads dataset partitioning it in year/month folder
+nyc_dataset = ds.dataset('nyc-taxi/', partitioning=["year", "month"])
+
+# Gets Database Connection
+con = duckdb.connect()
+
+query = con.execute("SELECT * FROM nyc")
```

Comment:
Wait, how does DuckDB know that "nyc" == "nyc_dataset"? Is there a line missing here?

Reply:
@pdet?

Reply:
Yes. It should be nyc_dataset